Modernize storage page - stop using bootstrap classes and use PF4 components instead

KKoukiou opened the pull request:
Since we have the cards design in many pages already, I felt that the storage page felt somehow awkward.
I feel now it's more consistent with the rest of the pages.

Now:
![Screen Shot 2020-08-26 at 23 46 44](https://user-images.githubusercontent.com/14921356/91363592-3fb83380-e7fd-11ea-80b2-c6fc0eb9ebdc.png)

Before:
![Screen Shot 2020-08-27 at 00 35 56](https://user-images.githubusercontent.com/14921356/91363629-51014000-e7fd-11ea-97bb-2279c21cf727.png)

andreasn reviewed (commented):
Going to do a UX review before I start digging into the CSS.
Great work in general! Very nice change!

Some smaller things:
The padding around the list here is different than the rendering on the Networking page now. I think I like the rendering on the Networking page better where the lines touch the edges.
![Screenshot from 2020-08-27 10-50-10](https://user-images.githubusercontent.com/209895/91422645-bbdf6500-e857-11ea-8ca3-b5fd1c6ff61d.png)

There is a line below the empty state in the NFS section.
![Screenshot from 2020-08-27 10-50-15](https://user-images.githubusercontent.com/209895/91422779-e92c1300-e857-11ea-8140-2c22294401b1.png)

You're probably a bit restrained by the padding the elements give you by default, so don't worry too much about this one, but there is a lot of indentation here. The hierarchy between the title "Storage logs" and the date is already clear by the different font sizes, so no real need for the indentation.
![Screenshot from 2020-08-27 10-51-57](https://user-images.githubusercontent.com/209895/91422823-fcd77980-e857-11ea-8425-94850f527ce6.png)

Physical volumes and logical volumes changed position all of a sudden. If we decide it's better with that order, I would still like to see it broken out
![Screenshot from 2020-08-27 11-21-08](https://user-images.githubusercontent.com/209895/91423199-84bd8380-e858-11ea-96c9-32ae11bd3400.png)

KKoukiou commented:
> Going to do a UX review before I start digging into the CSS.
> Great work in general! Very nice change!
> 
> Some smaller things:
> The padding around the list here is different than the rendering on the Networking page now. I think I like the rendering on the Networking page better where the lines touch the edges.

Should I remove the side padding from the cards in the side panel as well? (Devices, Drives etc). Otherwise these would be inconsistent.

> ![Screenshot from 2020-08-27 10-50-10](https://user-images.githubusercontent.com/209895/91422645-bbdf6500-e857-11ea-8ca3-b5fd1c6ff61d.png)
> 
> There is a line below the empty state in the NFS section.
> ![Screenshot from 2020-08-27 10-50-15](https://user-images.githubusercontent.com/209895/91422779-e92c1300-e857-11ea-8140-2c22294401b1.png)
> 
> You're probably a bit restrained by the padding the elements give you by default, so don't worry too much about this one, but there is a lot of indentation here. The hierarchy between the title "Storage logs" and the date is already clear by the different font sizes, so no real need for the indentation.
> ![Screenshot from 2020-08-27 10-51-57](https://user-images.githubusercontent.com/209895/91422823-fcd77980-e857-11ea-8425-94850f527ce6.png)
> 
> Physical volumes and logical volumes changed position all of a sudden. If we decide it's better with that order, I would still like to see it broken out
> ![Screenshot from 2020-08-27 11-21-08](https://user-images.githubusercontent.com/209895/91423199-84bd8380-e858-11ea-96c9-32ae11bd3400.png)

andreasn commented:
> Should I remove the side padding from the cards in the side panel as well? (Devices, Drives etc). Otherwise these would be inconsistent.

Yes, I think that would work better. Can't tell for sure before I see it, but leaning towards it being the right way.

garrett commented:
Note: Some of the indentation for the logs is due to having space reserved for an icon.

As for the rest, I'll look at this soon and adjust things.

Really happy to see this PR happen. Huge kudos for proactively porting the page over, @KKoukiou!

marusak commented:
On interface details there should be grey background and some spacing between graphs and the card? And empty state for ports?
![nogrey](https://user-images.githubusercontent.com/12330670/91717600-1ba87980-eb92-11ea-8ebd-66523a6f7214.png)

KKoukiou commented:
> On interface details there should be grey background and some spacing between graphs and the card? And empty state for ports?
> ![nogrey](https://user-images.githubusercontent.com/12330670/91717600-1ba87980-eb92-11ea-8ebd-66523a6f7214.png)

Yes, this has nothing to do with this PR though, it's a regression from:
commit 60571e3c599f5bc98a7a034204be965634427a45
Author: Katerina Koukiou <kkoukiou@redhat.com>
Date:   Thu Aug 6 17:13:34 2020 +0200

    network: Utilize cards and PF4 tables for showing the interfaces lists


I am working on the fix right now.

marusak commented:
https://github.com/cockpit-project/cockpit/issues/14534
Otherwise the code changes look good. Not approving for now due to ^ and since Garrett wants to have a look as well.

garrett commented:
Yep, compiling now. I'm looking forward to "kicking the tires" and seeing if anything needs some adjustments (aside from the content-not-filling-page issue).

martinpitt commented on pkg/storaged/storage.css:
```diff
@@ -16,6 +16,7 @@
  * You should have received a copy of the GNU Lesser General Public License
  * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
  */
+@import "../../node_modules/@patternfly/patternfly/components/Page/page.css";
```
OOI, why does that need to happen here? I thought when importing PF-React in overview.jsx it would bundle up the CSS inside the webpack? ISTR that this explicit "include in .css" is only necessary when using `pf-*` classes directly in .html files?

KKoukiou replied:
There are three components, including this one (Page), which need special handling because they use some PF global variables for breakpoints which we override, see

https://github.com/cockpit-project/cockpit/blob/master/webpack.config.js#L533
and
https://github.com/cockpit-project/cockpit/blob/master/pkg/lib/_global-variables.scss#L8

For these we forbid loading the auto imported CSS files so that we can manually import the SCSS so that our global variables overrides can take effect.

martinpitt replied:
Thanks!

martinpitt reviewed (commented):
Code-wise this looks great, I just have one minor question. :+1: Leaving the final yay/nay to @garrett though, I think he still found some issues (a11y and maybe others)

KKoukiou commented on pkg/storaged/devices.jsx:
```diff
@@ -88,7 +88,7 @@ class StoragePage extends React.Component {
             <>
                 <MultipathAlert client={client} />
-                <div hidden={!!detail}><Overview client={client} /></div>
+                <Overview client={client} hidden={!!detail} />
```
This looks wrong, hidden is not a property of the Overview component. I will test it tomorrow in practice.

garrett replied:
It's wrong to have an empty div wrapping the page. This is why it wasn't stretching properly. 

Hidden is a property of HTML elements, and it worked properly in testing.

garrett replied:
Having that div in the way was causing https://github.com/cockpit-project/cockpit/issues/14534

KKoukiou replied:
I just tried it out and indeed this breaks navigation to the sub pages.

I don't say that the div should be there, just the proper fix is:
```
{!detail && <Overview client={client} />}
```

Overview is a React component, that what you pass there are component properties, not html attributes.

The tests pass because your change makes both pages main and details both appear in the html.

KKoukiou replied:
In addition there are in the CSS the height: 100% rules which were trying to address the same issue, and these can be removed now.

KKoukiou commented on pkg/storaged/overview.jsx:
```diff
@@ -49,7 +49,7 @@ export class Overview extends React.Component {
                             <Card>
                                 <CardBody>
-                                    <StoragePlots client={client} onHover={(dev) => this.setState({ highlight: dev })} />
+                                    <StoragePlots client={client} className="storage-graphs" onHover={(dev) => this.setState({ highlight: dev })} />
```
This just adds a class which is not corresponding to some rule.
Maybe you forgot something?

KKoukiou commented:
I pushed a fix for the broken navigation as discussed above in the review, removed the commit about graphs padding that was not introducing any functionality and squashed the commits that were not independent.

KKoukiou commented:
@garrett I think you forgot to downsize the font of the Logs header. I will post a commit for that.

garrett commented:
Added a few log fixes (class, hiding headings without content), squashed and force-pushed

garrett commented:
Page gets a little weird when at a small size. Working on that now.

garrett commented:
I fixed the layout on mobile and desktop, so that the page scrolls, not the grid. And I [normalized the empty states](https://github.com/cockpit-project/cockpit/pull/14513/commits/44ef18b585565c339639a3716dd1a74b18a6b06e#diff-40973a11ec41f70d0e3cd499cfaca6e3R76-R77) (in CSS).

garrett reviewed (approved):
I think this is good to merge now. :+1:

KKoukiou commented:
> I think this is good to merge now. +1

Something broke one test from storage page today, https://logs-https-cockpit.apps.ci.centos.org/logs/pull-14513-20200901-112759-32f524db-fedora-33/log.html#241-2

marusak reviewed (approved):
Thank you @KKoukiou !
Code looks good, UI as well. Tests and @garrett agree so lets get this in into today's release :)
